rivasj/incubator-trafficcontrol
# Pattern-based consistent hashing feature for TR/TO/TP

#### What does this PR do?
Pattern-based consistent hashing is a new feature that allows the user to influence cache selection (and delivery service selection for steering ds's) by adjusting the request path before it is passed to the consistent hasher in Traffic Router by applying a regex group matching. This is accomplished via a new delivery service field on HTTP and Steering delivery services called "Consistent Hash Regex" that is propagated to TR through the CRConfig.

Traffic Router parses the new optional regex field from the CRConfig and uses it during cache selection (HTTP delivery services) or delivery service selection (Steering delivery services).

For HTTP delivery services:
- consistent hash regex is applied to request path to extract pieces of the original request path to 'normalize' a string before passing to the consistent hasher for cache selection

For Steering delivery services:
- consistent hash regex is applied to request path to extract pieces of the original request path to 'normalize' a string before passing to the consistent hasher for delivery service selection
- the selected (target) delivery service then inherits the consistent hash regex from the parent/steering delivery service and uses it as an HTTP delivery service does

Additionally, I've included a Traffic Portal "Test Tool" so that users can test a regex against a request path to determine exactly what will be used by Traffic Router's consistent hasher. There is a link to the test tool underneath the form field in Traffic Portal.

#### Which TC components are affected by this PR?


#### What is the best way to verify this PR?
- run Traffic Router unit and integration tests
- run Traffic Ops unit and integration tests
- install the components (TP/TO/TR) and verify the "consistentHashRegex" field makes it into the CrConfig
- verify the TP "Test Tool" works as expected by providing various inputs

#### Check all that apply
